source/extensions/network/dns_resolver/getaddrinfo/getaddrinfo.cc

```diff
@@ -0,0 +1,217 @@
+
+#include "envoy/extensions/network/dns_resolver/getaddrinfo/v3/getaddrinfo_dns_resolver.pb.h"
+#include "envoy/network/dns_resolver.h"
+
+#include "source/common/api/os_sys_calls_impl.h"
+#include "source/common/network/address_impl.h"
+
+namespace Envoy {
+namespace Network {
+
+// This resolver uses getaddrinfo() on a dedicated resolution thread. Thus, it is only suitable
+// currently for relatively low rate resolutions. In the future a thread pool could be added if
+// desired.
+class GetAddrInfoDnsResolver : public DnsResolver, public Logger::Loggable<Logger::Id::dns> {
+public:
+  GetAddrInfoDnsResolver(Event::Dispatcher& dispatcher, Api::Api& api)
+      : dispatcher_(dispatcher),
+        resolver_thread_(api.threadFactory().createThread([this] { resolveThreadRoutine(); })) {}
+
+  ~GetAddrInfoDnsResolver() override {
+    {
+      absl::MutexLock guard(&mutex_);
+      shutting_down_ = true;
+    }
+
+    resolver_thread_->join();
+  }
+
+  // DnsResolver
+  ActiveDnsQuery* resolve(const std::string& dns_name, DnsLookupFamily dns_lookup_family,
+                          ResolveCb callback) override {
+    ENVOY_LOG(debug, "adding new query [{}] to pending queries", dns_name);
+    auto new_query = std::make_unique<PendingQuery>(dns_name, dns_lookup_family, callback);
+    absl::MutexLock guard(&mutex_);
+    pending_queries_.emplace_back(std::move(new_query));
+    return pending_queries_.back().get();
+  }
+
+  void resetNetworking() override {}
+
+private:
+  class PendingQuery : public ActiveDnsQuery {
+  public:
+    PendingQuery(const std::string& dns_name, DnsLookupFamily dns_lookup_family, ResolveCb callback)
+        : dns_name_(dns_name), dns_lookup_family_(dns_lookup_family), callback_(callback) {}
+
+    void cancel(CancelReason) override {
+      ENVOY_LOG(debug, "cancelling query [{}]", dns_name_);
+      cancelled_ = true;
+    }
+
+    const std::string dns_name_;
+    const DnsLookupFamily dns_lookup_family_;
+    ResolveCb callback_;
+    bool cancelled_{false};
+  };
+  // Must be a shared_ptr for passing around via post.
+  using PendingQuerySharedPtr = std::shared_ptr<PendingQuery>;
+
+  // Parse a GAI response and determine the final address list.
+  std::pair<ResolutionStatus, std::list<DnsResponse>>
+  processGaiResponse(const PendingQuery& query, addrinfo* addrinfo_result) {
+    std::list<DnsResponse> v4_results;
+    std::list<DnsResponse> v6_results;
+    for (auto ai = addrinfo_result; ai != nullptr; ai = ai->ai_next) {
+      if (ai->ai_family == AF_INET) {
+        sockaddr_in address;
+        memset(&address, 0, sizeof(address));
+        address.sin_family = AF_INET;
+        address.sin_port = 0;
+        address.sin_addr = reinterpret_cast<sockaddr_in*>(ai->ai_addr)->sin_addr;
+
+        v4_results.emplace_back(DnsResponse(std::make_shared<const Address::Ipv4Instance>(&address),
+                                            std::chrono::seconds(60)));
```

Comment:
where does 60 come from?

Reply:
It seemed like a reasonable default given that this API doesn't support getting TTL. Would you prefer a different default? I will document it and I guess we could make it configurable later if anyone cares.

Reply:
yeah definitely want comments.  If this is the default you want for Lyft I'd say you can skip the knob for now.